west_commands: runners: openocd: clear DHCSR after flashing completes

ioannisg opened the pull request:
Fixes #32984 

openOCD does not clear DHCSR.DEBUGEN bit after flashing
completes, effectively leaving the Cortex-M processor in
debug mode. Manually clean DHCSR becore exiting the flash
operation, to work around the problem.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
This is just a trial for now - do not merge, since the fix should only be applicable for Cortex-M.
@erwango could you check if this would work for openOCD-flashed targets?

mbolivar-nordic commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -168,6 +168,7 @@ def do_flash(self, **kwargs):
                ['-c', self.verify_cmd + ' ' + hex_name] +
                post_verify_cmd +
                ['-c', 'reset run',
+                '-c', 'mww 0xE000EDF0 0xA05F0000', #32984
```
Thanks for this. No comments on the register writes themselves, but thoughts on getting this out of draft:

To abstract this properly, I think we need to either:

1. move these into one of the existing command line hooks (e.g. the one that ends up in `post_verify_cmd`), or
2. create a new command line hook

After that, we can either:

1. make a `boards/common/openocd-cortex-m.board.cmake` file that sets this up appropriately, and include it from all the right board.cmake files (not forgetting to update `openocd-nrf5.board.cmake` too), or
2. modify openocd.board.cmake to check Kconfig variables for Cortex-M and do it automatically the way it already looks at `CONFIG_X86` and `CONFIG_ARC` to decide if it uses `load_image` or `flash write_image erase`

Based on the way you've written this PR, it looks like you'll need to add a new command line option since this doesn't appear next to any of the existing `_cmd` lists here or below. I think modifying `openocd.board.cmake` instead of creating a new file will be better and easier.

mbolivar-nordic replied:
I'm also kind of surprised nobody noticed this yet. Is there a standard openocd command for this that doesn't require touching the register directly?

galak replied:
> I'm also kind of surprised nobody noticed this yet. Is there a standard openocd command for this that doesn't require touching the register directly?

Not that we are aware of.  There's nothing in the code that is obvious.

ioannisg replied:
I sent a patch request upstream: http://openocd.zylin.com/#/c/6093/1 to change the default mode.

ioannisg committed:
scripts: openocd.py: add option to clear halt debug enable state

Add a new option to the openocd.py script, for clearing
the halt debug enable state.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: common: add cortex-m specific command for clearing DHCSR

Introduce a command in openocd.board.cmake for cortex-m
targets, which clears the halt debug enable bit flag, upon
openocd exit. This is required since openocd does not clear
the state itself when it shuts down.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

mbolivar-nordic commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -167,8 +178,9 @@ def do_flash(self, **kwargs):
-               ['-c', 'reset run',
-                '-c', 'shutdown'])
+               ['-c', 'reset run'] +
+               clear_halt_debug_en_cmd +
```
This name is confusing mechanism with policy. Please rename it to `post_reset_cmd` instead.

ioannisg replied:
Yeap, that's fine.

mbolivar-nordic commented on boards/common/openocd.board.cmake:
```diff
@@ -16,7 +16,13 @@ endif()
 set(OPENOCD_CMD_LOAD_DEFAULT "${OPENOCD_FLASH}")
 set(OPENOCD_CMD_VERIFY_DEFAULT "verify_image")
 
+# Cortex-M-specific halt debug enable bit clearing
```
Can you add some more comments here, since this looks "weird"? Maybe just linking to the openocd PR you opened would be good enough.

ioannisg replied:
I'll comment properly.

mbolivar-nordic reviewed (changes requested):
Thanks @ioannisg . I read the discussion in upstream openocd and it looks like you really found something new here. A couple of minor requests and then I am +1 from the runner perspective.

However, I want to confirm with @erwango and other openocd users that we really want this to be the default behavior before merging, since the openocd maintainers named some use cases where they didn't think it made a sensible default.

ioannisg commented:
> However, I want to confirm with @erwango and other openocd users that we really want this to be the default behavior before merging, since the openocd maintainers named some use cases where they didn't think it made a sensible default.

Fine. I 've, also added the HW block label, so this is tested before it (ever) gets merged.

My understanding is that "west flash" shall only program the binary and reset the core to a state as close as possible to the (cold) reset state. Since the internal runner implementation does not guarantee this, we are left with the option to do this in the runner.

erwango commented:
@ioannisg is this PR still required if http://openocd.zylin.com/#/c/6093/1 gets merged ?

ioannisg committed:
boards: common: pyocd: cortex-m specific command for clearing DHCSR

Introduce a command in pyocd.board.cmake for ARM Cortex-m
targets, which clears the halt debug enable bit flag, after
flashing is complete and before pyocd exit. This is required
since pyocd regular flashing command does not clear the state
itself when it completes, nor when pyocd session shuts down.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
scripts: pyocd.py: add option to clear halt debug enable state

Add a new option to the pyocd.py script, for clearing
the halt debug enable state. The option results in a
command that is called after the flashing is complete.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

erwango commented:
One consideration on current behavior is that having DHCSR set after flash is that it prevents target entering sleep states when testing PM feature.
In case DHCSR is cleared and PM sleep state is entered, flashing using openocd is not possible anymore and another flashing method shoud be used.
So, having DHCSR set after flash:
- is bad for measurements... BUT
- allows PM testing in CI with existing flashing methods.

mbolivar-nordic commented on scripts/west_commands/runners/pyocd.py:
```diff
@@ -158,6 +163,24 @@ def flash(self, **kwargs):
         self.logger.info('Flashing file: {}'.format(fname))
         self.check_call(cmd)
 
+        if self.post_flash_cmd is not []:
```
This is not correct usage. In Python, `foo is bar` and `foo is not bar` basically work like pointer comparison in C.

Example:

```py
>>> [] is []
False
```

This should be `if self.post_flash_cmd:` instead, because an empty list is boolean false.

mbolivar-nordic commented on boards/common/pyocd.board.cmake:
```diff
@@ -1,5 +1,17 @@
 board_set_debugger_ifnset(pyocd)
-board_finalize_runner_args(pyocd "--dt-flash=y")
+board_finalize_runner_args(pyocd "--dt-flash=y"
+    --post-flash "${PYOCD_CMD_HALT_DEBUG_EN_CLEAR}")
```
This is reading from an uninitialized variable on non-Cortex-M

erwango reviewed (changes requested):
Nacking until we have an alternative viable option to allow Power Management testing in CI runs.

As previously stated, current openocd behavior regarding debug enable bit allows to flash target after sleep mode code has been exercised, while the proposed change prevents this.

galak commented:
@ioannisg any updates here?  We should set the default of the feature to disabled until we resolve a number of these issues.
